### PR TITLE
Center Doubleclick Ads

### LINKED ISSUE #4693: When amp-ad is configured with layout=responsive, always center ad in iframe
When publishers use layout=responsive for amp-ad and the response is a fixed pixel creative, the creative ends up being left and top aligned. 
It should be vertically and horizontally centered regardless of the response.

You can see this in:
screen shot 2016-08-24 at 4 47 06 pm

Live article : http://www.theatlantic.com/amp/article/497159/
Code snippet : http://output.jsbin.com/fehayafoso

## What browsers are affected?

I tested this in Safari on an iphone 6.

## Which AMP version is affected?

It was always this way.

### PR DESCRIPTION
**Pull request by mlb7687**
This adds the layout to window.context so that we can use it in cross domain ad iframe.
We will need this ability to support multi-size ads and pad as well.
For DoubleClick, if we see a responsive ad, pad and size things appropriately.

Fixes #4693.

### CODE REVIEW

File: ads/google/doubleclick.js [hunk 1]
```diff
@@ -79,6 +79,15 @@ function doubleClickWithGpt(global, data, gladeExperiment) {
     parseInt(data.overrideHeight || data.height, 10),
   ]];
 
+  // If the ad layout is responsive, center the fixed size ad in the container.
+  if (global.context.layout == 'responsive') {
+    const container = global.document.querySelector('#c');
+    container.style.top = '50%';
+    container.style.left = '50%';
+    container.style.margin = dimensions[0][1]/-2 + 'px 0 0 ' +
```
Comment (jridgewell) [inline ads/google/doubleclick.js]: Using `transform: translate(-50%, -50%)` skips the halving math.

File: ads/google/doubleclick.js [hunk 2]
```diff
@@ -190,14 +199,25 @@ function doubleClickWithGlade(global, data, gladeExperiment) {
...
-  slot.setAttribute('data-request-width', requestWidth);
+  if (global.context.layout == 'responsive') {
+    // If the slot is responsive, we can have a different size ad come back
+    // than the container size, so explicitly set height and width, then center
+    // in the container.
+    slot.setAttribute('height', requestHeight);
+    slot.setAttribute('width', requestWidth);
+    slot.style.top = '50%';
+    slot.style.left = '50%';
+    slot.style.margin = requestHeight/-2 + 'px 0 0 ' + requestWidth/-2 + 'px';
```
Comment (jridgewell) [inline ads/google/doubleclick.js]: ditto.

File: src/3p-frame.js [hunk 3]
```diff
@@ -81,6 +81,7 @@ function getFrameAttributes(parentWindow, element, opt_type) {
     location: {
       href: locationHref,
     },
+    layout: element.getAttribute('layout'),
```
Comment (jridgewell) [inline src/3p-frame.js]: @dvoytenko: Should we be exposing our layout types to ads?

**Comment by jridgewell** [comment]
Code LGTM. I'm not sure we should be exposing out layout types, though.

File: ads/google/doubleclick.js [hunk 4]
```diff
@@ -79,6 +79,15 @@ function doubleClickWithGpt(global, data, gladeExperiment) {
     parseInt(data.overrideHeight || data.height, 10),
   ]];
 
+  // If the ad layout is responsive, center the fixed size ad in the container.
+  if (global.context.layout == 'responsive') {
```
Comment (jasti) [inline ads/google/doubleclick.js]: What's the downside of centering for all layouts?

Comment (mlb7687) [inline ads/google/doubleclick.js]: Good point, centering in an ad frame the same size won't do anything. Do you foresee needing to know the layout for responsive multi-size use case? If not, I can get rid of it from window.context

Comment (jasti) [inline ads/google/doubleclick.js]: Don't think so. Even in the multi-size case, I think we'll always want to center the ad response.

**Comment by mlb7687** [comment]
Fixed. PTAL.

File: ads/google/doubleclick.js [hunk 5]
```diff
@@ -79,6 +79,14 @@ function doubleClickWithGpt(global, data, gladeExperiment) {
     parseInt(data.overrideHeight || data.height, 10),
   ]];
 
+  // Center the ad in the container.
+  const container = global.document.querySelector('#c');
```
Comment (tdrl) [inline ads/google/doubleclick.js]: I assume that when you get here, this selector is guaranteed to return a non-null element?

Comment (mlb7687) [inline ads/google/doubleclick.js]: We already make this assumption for every GPT Light request which is running at 90% (see a few lines up from changes there).
I think this is a fairly safe assumption if people aren't explicitly removing this from their remote.html files. Perhaps the AMP folks would be better at saying what the policy here should be.
If we decide to consider this unsafe, we should change this for both GPT and GPT Light.

**Comment by jridgewell** [comment]
LGTM.

File: ads/google/doubleclick.js [hunk 6]
```diff
@@ -79,6 +79,14 @@ function doubleClickWithGpt(global, data, gladeExperiment) {
...
+  container.style.top = '50%';
+  container.style.left = '50%';
+  container.style.bottom = '';
+  container.style.right = '';
+  container.style.transform = 'translate(-50%, -50%)';
```
Comment (dvoytenko) [inline ads/google/doubleclick.js]: Is `#c` positioned absolutely?

File: ads/google/doubleclick.js [hunk 7]
```diff
@@ -189,15 +197,14 @@ function doubleClickWithGlade(global, data, gladeExperiment) {
...
-  // The ad container should simply fill the amp-ad iframe, but we still
-  // need to request a specific size from the ad server.
-  // The ad container size will be relative to the amp-iframe, so if the
-  // latter changes the ad container will match it.
-  slot.setAttribute('width', 'fill');
-  slot.setAttribute('height', 'fill');
-  slot.setAttribute('data-request-height', requestHeight);
-  slot.setAttribute('data-request-width', requestWidth);
+  // Center the ad in the container.
+  slot.setAttribute('height', requestHeight);
```
Comment (dvoytenko) [inline ads/google/doubleclick.js]: Ditto here?

File: ads/google/doubleclick.js (hunk 6 above)
Comment (mlb7687) [inline ads/google/doubleclick.js]: This is the div from frame.html/remote.html set to position absolute with positions 0,0,0,0. Same for other comment.
https://github.com/ampproject/amphtml/blob/master/3p/remote.html

Comment (dvoytenko) [inline ads/google/doubleclick.js]: Sounds good. thanks!

**Comment by mlb7687** [comment]
This is ready to be merged whenever you guys get a chance

**Pull request merged**